File: plugins-community/install-helper/src/core/registry.ts
// ----------------------------------------------------------------------------------------------------------
// Copyright (c) 2026 Huawei Technologies Co., Ltd.
// This program is free software, you can redistribute it and/or modify it under the terms and conditions of
// CANN Open Software License Agreement Version 2.0 (the "License").
// Please refer to the License for details. You may not use this file except in compliance with the License.
// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
// See LICENSE in the root of the software repository for the full text of the License.
// ----------------------------------------------------------------------------------------------------------

import { readdirSync, existsSync } from "fs";
import { join } from "path";
import type { PluginEntry } from "../types/index.js";
import { isDirectory } from "../utils/fs-helpers.js";
import embeddedPlugins from "../embedded-plugins.json" with { type: "json" };

function loadPlugins(): PluginEntry[] {
  const plugins = [...embeddedPlugins] as PluginEntry[];
  plugins.sort((a, b) =>
    a.displayName < b.displayName ? -1 : a.displayName > b.displayName ? 1 : 0
  );
  return plugins;
}

export let PLUGIN_REGISTRY: PluginEntry[] = loadPlugins();

const PLUGIN_DIRS = ["plugins-official", "plugins-community"];

export function mergeDynamicPlugins(repoPath: string): void {
  if (!existsSync(repoPath)) return;

  const existingIds = new Set(PLUGIN_REGISTRY.map((p) => p.id));
  const dynamic: PluginEntry[] = [];

  for (const pluginDir of PLUGIN_DIRS) {
    const parent = join(repoPath, pluginDir);
    if (!existsSync(parent)) continue;

    let entries: string[] = [];
    try {
      entries = readdirSync(parent);
    } catch {
      continue;
    }

    for (const entry of entries) {
      const pluginPath = join(parent, entry);
      if (!isDirectory(pluginPath)) continue;
      if (entry === "install-helper") continue;

      const initSh = join(pluginPath, "init.sh");
      if (!existsSync(initSh)) continue;

      const pluginId = entry;
      if (existingIds.has(pluginId)) continue;

      dynamic.push({
        id: pluginId,
        dir: `${pluginDir}/${entry}`,
        displayName: entry,
        script: "init.sh",
        aliases: [],
        skills: 0,
        agents: 0,
        description: "",
      });
      existingIds.add(pluginId);
    }
  }

  if (dynamic.length > 0) {
    PLUGIN_REGISTRY = [...PLUGIN_REGISTRY, ...dynamic].sort((a, b) =>
      a.displayName < b.displayName ? -1 : a.displayName > b.displayName ? 1 : 0
    );
  }
}

export function findPlugin(query: string): PluginEntry | undefined {
  const normalized = query.toLowerCase().trim();

  const exactMatch = PLUGIN_REGISTRY.find(
    (p) => p.id === normalized || p.id === query.trim()
  );
  if (exactMatch) return exactMatch;

  const aliasMatch = PLUGIN_REGISTRY.find((p) =>
    p.aliases.some((a) => a === normalized)
  );
  if (aliasMatch) return aliasMatch;

  const prefixMatches = PLUGIN_REGISTRY.filter((p) => p.id.startsWith(normalized));
  if (prefixMatches.length === 1) return prefixMatches[0];

  return undefined;
}

export function getAllPlugins(): PluginEntry[] {
  return PLUGIN_REGISTRY;
}

export function getPluginById(id: string): PluginEntry | undefined {
  return PLUGIN_REGISTRY.find((p) => p.id === id);
}

File: plugins-community/ascendc-port-orchestrator/engine/src/scripts/patches/build_capabilities/registry.py
# ----------------------------------------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software; you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------------------------------------

"""Capability registry — name -> factory((BuildContext) -> CMakeContribution).

Named capabilities (e.g. `kfc_bootstrap`, `il_lowering`) register here and are
resolved from a kernel's declared `capabilities` array. A declared name that does
not resolve FAILS LOUDLY (design §2A) — get() raises, never silently skips.

Baseline accretions (DEBT-20/20.1, DEBT-110) are NOT registered here: they always
apply and live in build_capabilities.baseline.
"""
from __future__ import annotations

from typing import Callable

from .context import BuildContext
from .contribution import CMakeContribution

Factory = Callable[[BuildContext], CMakeContribution]


class Registry:
    def __init__(self) -> None:
        self._factories: dict[str, Factory] = {}

    def __contains__(self, name: str) -> bool:
        return name in self._factories

    def register(self, name: str, factory: Factory) -> None:
        if name in self._factories:
            raise ValueError(f"build capability '{name}' already registered")
        self._factories[name] = factory

    def get(self, name: str) -> Factory:
        try:
            return self._factories[name]
        except KeyError:
            raise KeyError(
                f"unknown build capability '{name}'; "
                f"registered: {sorted(self._factories)}"
            ) from None

    def names(self) -> list[str]:
        return sorted(self._factories)


# Process-wide registry. Named capability modules register on import.
registry = Registry()
